homeassistant/components/geo_location/__init__.py

```diff
@@ -0,0 +1,70 @@
+"""
+Geo Location component.
+
+This component covers platforms that deal with external events that contain
+a geo location related to the installed HA instance.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/geo_location/
+"""
+import logging
+from datetime import timedelta
+from typing import Optional
+
+from homeassistant.const import ATTR_LATITUDE, ATTR_LONGITUDE
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity_component import EntityComponent
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_DISTANCE = 'distance'
+DOMAIN = 'geo_location'
+ENTITY_ID_FORMAT = DOMAIN + '.{}'
+GROUP_NAME_ALL_EVENTS = 'All Geo Location Events'
+SCAN_INTERVAL = timedelta(seconds=60)
+
+
+async def async_setup(hass, config):
+    """Setup this component."""
+    component = EntityComponent(
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL, GROUP_NAME_ALL_EVENTS)
+    await component.async_setup(config)
+    return True
+
+
+class GeoLocationEvent(Entity):
+    """This represents an external event with an associated geo location."""
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        if self.distance is not None:
+            return round(self.distance, 1)
+        return None
+
+    @property
+    def distance(self) -> Optional[float]:
+        """Return distance value of this external event."""
+        return None
+
+    @property
+    def latitude(self):
+        """Return latitude value of this external event."""
+        return None
+
+    @property
+    def longitude(self):
+        """Return longitude value of this external event."""
+        return None
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes of this external event."""
+        data = {}
+        if self.distance is not None:
+            data[ATTR_DISTANCE] = self.distance
```

Comment:
The distance is already available as `state` if one wants to use it for automations.

Reply:
The reason why there is a separate distance property is because the distance is probably a useful default state for most use-cases in this component. However, depending on the nature of the external event, other attribute may be more important and used as the state of an entity. Example: If the external event is an earthquake, the magnitude becomes much more relevant.

In a separate branch I have already built a GeoRSS platform where the user can choose in the configuration what state the entities should have.

Reply:
Or, are you saying that I should remove the distance from the `state_attributes` because it is already available as `state`, and if a platform decides to expose something other than the distance as `state` then this platform could add the distance as an attribute?

Reply:
Alright, and how does that relate to for example the `sensor` component. There are a lot of platforms in that component that return plenty of different things - temperature, humidity, luminance, number of Fedex deliveries, ...?

Reply:
The sensor component doesn't have a specific state definition. All other platforms have a specific state definition.